README.md
# Splash

Splash is a [g-speak](http://platform.oblong.com/) program that transforms the Leap Motion's relative coordinates to absolute coordinates.  Splash streams its data output as a series of messages into a pool named **leap**.  Any number of other programs can listen to that pool and get the data.

(Pools are part of the g-speak's message-passing system known as **Plasma**.  It's a lighweight publish/subscribe system that scratches some of the same itches as OSC or 0MQ's pubsub.  More details about **Plasma** [can be found here](https://platform.oblong.com/learning/g-speak/structure/libplasma/).)

Splash is a command-line utility -- it runs in the terminal only.

Splash reads Leap data using the Leap SDK (which you must have already installed).  Splash also assumes you have already installed g-speak, an SDK from Oblong for creating spatial, gestural, multi-modal, and multi-machine applications. [Contact Us](mailto:platform-support@oblong.com) for licensing information.

To learn more about how to set up a g-speak program to work with Splash and Leap Motion input, see [this tutorial on the Platform site](http://greenhouse.oblong.com/learning/hardware_leap.html).

## Dependencies

- g-speak platform -- a proprietary SDK provided by Oblong Industries. Contact [Oblong Industries, Inc.](mailto:platform-support@oblong.com)
- [LeapSDK](https://developer.leapmotion.com/documentation/cpp/index.html)
- [CMake](https://cmake.org/)
  - Mac OS X using [brew](http://brew.sh/)

    $ brew install cmake

  - Ubuntu

    $ sudo apt-get install cmake


## Building

Make sure to set `LEAPSDK_HOME` to the directory where your Leap SDK lives.  Then build (we use our internal build utility, [obi](https://github.com/Oblong/obi.git) which works off of cmake).

    # With obi
    $ export LEAPSDK_HOME=/path/to/LeapSDK
    $ obi build

or

    # Without obi
    $ export LEAPSDK_HOME=/path/to/LeapSDK
    $ mkdir build
    $ cd build
    $ cmake ..
    $ make

### Leap SDK version

To support the [`POLICY_BACKGROUND_FRAMES`](https://developer.leapmotion.com/articles/testing-background-leap-applications) flag in the Leap Motion controller, `splash` requires at least version 0.7.9 of the Leap SDK.


## Running

### Before running splash

1. Connect a Leap device
2. Verify that the Leap system is working (e.g. with Leap's software).

### Actually running splash

The Leap shared library must be on your `LD_LIBRARY_PATH` (or `DYLD_LIBRARY_PATH` if you're using a Mac) before you start splash.  e.g.:

    # With obi
    $ obi go

or

    # Without obi
    $ DYLD_LIBRARY_PATH=/path/to/LeapSDK/lib
    $ build/splash

### To check that it's working

To see the output from `splash` in a human-readable format, use the peek command at the terminal: `peek leap`.  You should see data scrolling by quickly, representing the messages that splash is depositing into the **leap** pool; when you hover a hand in view of the Leap, the data should resemble the output shown below in the **Output Format** section.


### Arguments

Optionally, `splash` takes the fully qualified path to a simple YAML settings file that describes the physical location of the Leap sensor.  The file format is as follows:

    %YAML 1.1
    %TAG ! tag:oblong.com,2009:slaw/
    --- !protein
    descrips: # ignored
    ingests:
      leap:
        cent: [0, 0, 0]   # required, center of the leap device in absolute coordinates
        norm: [0, 1, 0]   # required, normal vector from the face of the leap device
        over: [1, 0, 0]   # required, vector pointing lengthwise along the device
        provenance: "leap-whatever"  # optional, defaults to "leap-$HOSTNAME"

'norm' is usually 'up' along the Y axis, because the Leap device is usually lying on a flat surface.

'over' is usually to the user's right (so it's positive in the X axis).

If a settings file is not supplied on the command line, `splash` will try to find a 'leap' section in the ingests of the `/etc/oblong/screen.protein` file.  (This file would have been installed by the Greenhouse installer.)

If a 'leap' section isn't found in screen.protein, `splash` will just assume that your Leap is 500mm in front and 200mm below the center of the "main" screen described in your `/etc/oblong/screen.protein` file.

If a `screen.protein` file can't be found, or if it does not contain a screen called "main", `splash` will warn you of as much and will pass the leap's data through in the native (relative) coordinates supplied by the Leap SDK.

For more information on how Greenhouse works with 3D space refer to this [Spatial Considerations tutorial](http://greenhouse.oblong.com/learning/spatial.html).

### Environment

By default, `splash` deposits its output (a series of messages which in Oblong parlance are referred to as *proteins*') into a pool called "leap".  Any number of Greenhouse programs on the same machine can listen to (or "particpate in") that pool.

If there are Greenhouse programs on a different machine that would like to listen, they can refer to the pool as "tcp://your-host-name/leap".

You can optionally specify a different pool to use via the `LEAP_POOL` environment variable.
Note that this pool must exist before you run `splash.`

    $ p-create my_new_leap_pool
    $ LEAP_POOL=my_new_leap_pool ./splash


## Output Format

To see the output from `splash` in a human-readable format, use the peek command.
`peek leap` (or the name of some other pool you specified) at the command line.

`splash` produces proteins of the following format:

    descrips:
    - greenhouse
    - leap
    - 0.7.9 # The version of the Leap Motion API
    ingests:
      leap:
        orig: v3float64 # the center of the Leap device itself
        norm: v3float64 # the leap's normal vector, usually [0,1,0]
        over: v3float64 # the direction to the leap's right, usually [1,0,0]
        prov: Str # the provenance associated with the leap
      frame:
        ts: int64 # Frame timestamp
        id: int64 # Frame ID
        hands: # a list of hands
        - id: int64
          dir: v3float64 # the direction the hand is pointing in
          plmnrm: v3float64 # palm normal vector
          plmpos: v3float64 # palm position
          plmvel: v3float64 # palm velocity
          center: v3float64 # Center of a sphere fit to the hand
          radius: float64 # radius of a sphere fit to the hand
          orig: v3float64 # palm position, ob-named
          thru: v3float64 # orig + dir
        pntrs: # a list of pointers (fingers, tools, etc.)
        - id: int64
          dir: v3float64 # the direction the pointer is pointing in
          hand: int64 # the ID of the hand the pointer is associated with
          isfngr: bool # Does the leap believe that this pointer is a finger?
          istool: bool # Does the leap believe that this pointer is a tool?
          length: float64 # the length of the pointer
          t-pos: v3float64 # Tip position
          t-vel: v3float64 # Tip velocity
          width: float64 # Width of the pointer
          orig: v3float64 # tip position again, ob-named
          thru: v3float64 # orig + dir * length
        gests: # A list of gestures
        - id: int64
          dur: int64 # How long has the gesture been active, microseconds.
          dursec: float64 # How long has the gesture been active, seconds.
          hands: # a list of hand ids associated with the gesture
          - int64
          - int64
          pntrs: # a list of pointer ids associated with the gesture
          - int64
          - int64
          state: Str # one of "start", "update", "stop" or "invalid"
          type: Str # one of "circle", "swipe", "s-tap", "k-tap" or "invalid"
          point: int64 # Id of pointer, only for circle, swipe, s-tap or k-tap
          ## The following occur in a "circle" gesture
          cent: v3float64 # Center of a "circle" gesture
          norm: v3float64 # Normal of a "circle" gesture
          prog: float64 # The number of times a finger has traversed a "circle"
          radius: float64 # The radius of a "circle" gesture
          ## The following occur in a "swipe" gesture
          dir: v3float64 # Direction of the swipe
          pos: v3float64 # Current position of the swipe
          speed: float64 # Speed of the swipe in mm/second
          start: v3float64 # The starting position of the swipe
          ## The following occur in s-tap (screen tap) and k-tap (key tap) gestures
          dir: v3float64 # Direction of the tap
          pos: v3float64 # Position of the tap
          prog: float64 # Always 1.0, for whatever reason



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## [0.0.1] - 1970-01-01

### Added
- development begins


src/main.cpp

/* (c)  oblong industries */

// # splash
//
// Help your Leap Motion sensor dive in to pools

#include <Leap.h>
#define WORKER
#include "libGreenhouse/Greenhouse.h"
#include <list>
#include <boost/thread/mutex.hpp>

#define MIN_CONFIDENCE 0.4

/// FrameWriter: A handy class to transform leap types into plasma types
class FrameWriter {
protected:
  Leap::Vector origin, normal, over;
  Leap::Matrix transform;
//  Str provenance;

  struct phalange {
    Str type;
    Vect loc, norm, over;
    bool occluded;
    phalange (const Str &t)
    { type = t;
      occluded = true;
    }
    Slaw toSlaw ()
    { return Slaw::Map ("type", type,
                        "loc", loc,
                        "norm", norm,
                        "over", over,
                        "occluded", occluded);
    }
  };

  /// Transform g-speak Vect into Leap Vector
  const Leap::Vector LV (Vect const& v) const
  { return Leap::Vector (v.x, v.y, v.z); }

  const Vect GV (Leap::Vector const& v) const
  { return Vect (v.x, v.y, v.z); }

  /// Takes a relative direction from Leap and returns it's absolute direction
  Leap::Vector const Direction (Leap::Vector const& v) const
  { return transform . transformDirection (v); }
  /// Takes a relative Location from Leap and returns it's absolute location
  Leap::Vector const Point (Leap::Vector const& v) const
  { return transform . transformPoint (v); }

  Slaw ProcessFingers (Leap::FingerList const &fl, const Vect &pNorm,
                       const Vect &pUp, Vect *aim, Str &gripe) const
  { char gstring [] = "_____";
    phalange phalanges [5] = { phalange ("THUMB"),
                               phalange ("INDEX"),
                               phalange ("MIDDLE"),
                               phalange ("RING"),
                               phalange ("PINKY") };
    // for (int i = 0)
    // Str types [5] = { "THUMB", "INDEX", "MIDDLE", "RING", "PINKY" };
    for (int i = fl.count () - 1  ;  i >= 0  ;  i--)
      { Leap::Finger f = fl[i];
        int t = f.type ();
        phalange p = phalanges [t];
        p.loc = GV (Point (f.stabilizedTipPosition ()));
        p.occluded = false;
        if (f.isExtended ())
          { if (t == 0)
              { Vect dir = GV (Direction (f.direction ()));
                gstring [4-t] = dir . AngleWith (pUp) < (0.3 * M_PI) ? '-' : '|';
              }
            else
              { gstring [4-t] = '|';
                if (t == 1 && aim)
                  { aim -> Set (GV (Direction (f.direction ())));
                    aim -> NormSelf ();
                  }
              }
          }
        else
          { if (t == 0)
              gstring [4-t] = '>';
            else
              { Vect dir = GV (Direction (f.direction ()));
                gstring [4-t] = dir . AngleWith (pNorm) < M_PI / 4.0 ? 'x' : '^';
              }
          }
        phalanges [t] = p;
      }
    gripe = Str () . Sprintf ("%s", gstring);
    return Slaw::List (phalanges [4].toSlaw (),
                       phalanges [3].toSlaw (),
                       phalanges [2].toSlaw (),
                       phalanges [1].toSlaw (),
                       phalanges [0].toSlaw ());
  }

  // Here we possibly incorrectly assume that the room's norm and over
  // are (0, 0, 1) and (1, 0, 0) respectively and that the person is
  // aligned with the leap and the room
  Str directionalGripe (Vect dir, bool leftish) const
  { float64 min_angle = 4.0 * M_PI;
    int32 min_dir = -1;

    Vect dirs [6] = { Vect (1, 0, 0), Vect (-1, 0, 0),
                      Vect (0, 1, 0), Vect (0, -1, 0),
                      Vect (0, 0, 1), Vect (0, 0, -1) };

    for (int i = 0  ;  i < 6  ;  i++)
      { float64 angle = dir . AngleWith (dirs [i]);
        if (angle < min_angle)
          { min_dir = i;
            min_angle = angle;
          }
      }

    switch (min_dir) {
      case 0: return leftish ? "-" : "+"; // lateral/medial
      case 1: return leftish ? "+" : "-"; // lateral/medial
      case 2: return "^"; // cranial
      case 3: return "v"; // caudal
      case 4: return "."; // posterior
      case 5: return "x"; // anterior
      default: return "_"; // unknown
    };

    return "_";
  }

  Slaw MissingHand (const Str &type) const
  { Vect v;
    return Slaw::Map ("gripe", "_____:__",
                      "type", type,
                      "loc", v,
                      "aim", v,
                      "back", Slaw::Map ("loc", v,
                                         "norm", v,
                                         "over", v,
                                         "occluded", true),
                      "fingers", Slaw::List (phalange ("PINKY").toSlaw (),
                                             phalange ("RING").toSlaw (),
                                             phalange ("MIDDLE").toSlaw (),
                                             phalange ("INDEX").toSlaw (),
                                             phalange ("THUMB").toSlaw ()));
  }

  Slaw ToSlaw (Leap::Hand const& h) const
  { Vect loc = GV (Point (h . stabilizedPalmPosition ()));
    Vect aim = GV (Direction (h . direction ())) . Norm ();
    Vect pnorm = GV (Direction (h . palmNormal ())) . Norm ();
    Vect norm = - pnorm;
    Vect up = norm . Cross (aim);
    Str type = h . isLeft () ? "LEFTISH" : "RIGHTISH";
    bool occ = h . confidence () >= MIN_CONFIDENCE;
    Str gripe = "_____:__";
    Slaw fingers = ProcessFingers (h . fingers (), pnorm, up, &aim, gripe);
    gripe = gripe + ":" + directionalGripe (pnorm, h . isLeft ()) +
            directionalGripe (aim, h . isLeft ());
    return Slaw::Map ("gripe", gripe,
                      "type", type,
                      "loc", loc,
                      "aim", aim,
                      "back", Slaw::Map ("loc", loc,
                                         "norm", norm,
                                         "over", up,
                                         "occluded", occ),
                      "fingers", fingers);
  }

  /// Creates a slaw that fully describes a Hand found by the Leap
  Slaw ToSlaw (Leap::HandList const& hl) const
  { if (hl . count () == 1)
      { Leap::Hand h = hl [0];
        Str missing = h . isLeft () ? "RIGHTISH" : "LEFTISH";
        return Slaw::List (ToSlaw (h),
                           MissingHand (missing));
      }
    else if (hl . count () == 0)
      { return Slaw::List (MissingHand ("RIGHTISH"), MissingHand ("LEFTISH"));
      }

    Slaw hands = Slaw::List ();
    for (int i = 0  ;  i < hl . count ()  ;  i++)
      hands = hands . ListAppend (ToSlaw (hl[i]));
    return hands;
  }

public:
  FrameWriter ()  :  transform (Leap::Matrix::identity ()) {}

  /// Sets up the location and orientation of the leap in order to properly
  /// transform relatively located leap events into absolutely located ones.
  void SetLocAndOrientation (Vect const& orig, Vect const& nrm, Vect const& ov)
  { origin = LV (orig);
    normal = LV (nrm);
    over = LV(ov);
    transform = Leap::Matrix (over, normal, over . cross (normal), origin);
  }

  /// Takes an event/frame generated from a leap and returns a Slaw that fully
  /// describes that event
  Slaw ToSlaw (Leap::Frame const& frame) const
  { return ToSlaw (frame . hands ());
  }

};

/// Callback handler for Leap events
template <typename Callback>
class SplashListener  :  public Leap::Listener {
protected:
  Callback &callback;
public:
  SplashListener (Callback &c)
    : callback (c) {}
  virtual void onFrame (const Leap::Controller &c)
  { callback . DepositGripes (c . frame ()); }
};

// Surely there's a way to get this from Leap's API...?
static const Str LEAP_VERSION = "0.7.9";
static const float64 DEFAULT_Z_DISTANCE = 500.0;
static const float64 DEFAULT_Y_DISTANCE = 200.0;

/// The main class type for Splash. It sets up the leap controller, the
/// callback handler for the controller and handles each frame using
/// the FrameWriter class to transform a leap frame into a g-speak protein
/// and deposits the protein into the designated output pool.
class Splash  :  public Thing {
private:
  Str const pool;
  SplashListener <Splash> listener;
  Leap::Controller controller;
  boost::mutex mutex;
  FrameWriter writer;
  std::list<Protein> to_deposit;
  bool stop_depositing;
  Str orig;

public:
  Splash (Str const& output_pool)
    :  Thing (),
       pool (output_pool),
       listener (*this),
       controller (listener),
       stop_depositing (false)
  { ParticipateInPool (pool);
    controller . setPolicyFlags (Leap::Controller::POLICY_BACKGROUND_FRAMES);
    orig = "leap-reader-v" + LEAP_VERSION;
  }

  virtual ~Splash ()
  { // Make sure we don't try to deposit any more proteins after
    // the application shuts down.
    { boost::mutex::scoped_lock lck (mutex);
      stop_depositing = true;
    }
    if (controller . isConnected ())
      controller . removeListener (listener);
  }

  /// The handler function for each Leap frame, it takes a leap frame,
  /// transforms it to a protein and deposits it to the leap pool
  void DepositGripes (Leap::Frame const& f)
  { int64 tim = controller.now ();//FatherTime::AbsoluteTime ();
    Protein p (Slaw::List ("gripeframe"),
               Slaw::Map ("origins", Slaw::List ("name", orig,
                                                 "clock", tim),
                          "time", tim,
                          "hands", writer . ToSlaw (f)));

    // The leap runs in its own thread.  Make sure that we don't try
    // to deposit two proteins at the same time.
    boost::mutex::scoped_lock lck (mutex);
    if (! stop_depositing)
      to_deposit . push_back (p);
  }

  virtual void Travail ()
  { std::list<Protein> copy;
    { boost::mutex::scoped_lock lck (mutex);
      std::swap (copy, to_deposit);
    }
    for (Protein const& p : copy)
      Deposit (p, pool);
  }

  /// Reads in a configuration protein to set up the spatial layout of
  /// the leap
  bool Configure (Protein const& p)
  { Slaw ing = p . Ingests ();
    Vect origin, normal, over;
    Str prov;
    Slaw leap = ing . MapFind ("leap");
    // if (leap . MapFind ("provenance") . Into (prov))
    //   writer . SetProvenance (prov);

    if (leap . MapFind ("cent") . Into (origin) &&
        leap . MapFind ("norm") . Into (normal) &&
        leap . MapFind ("over") . Into (over))
      { writer . SetLocAndOrientation (origin, normal, over);
        return true;
      }

    // If there's no explicit Leap information, infer it from
    // the screen protein, using the ``main'' screen.  We'll
    // assume that the leap is about 50cm back and 20 cm down
    // from the center of the screen and is pointing up.
    Slaw screen = ing . MapFind ("screens") . MapFind ("main");
    if (screen . MapFind ("cent") . Into (origin) &&
        screen . MapFind ("over") . Into (over))
      { normal = Vect (0, 1, 0);
        origin -= DEFAULT_Y_DISTANCE * normal . Norm ()
          + DEFAULT_Z_DISTANCE * normal . Cross (over) . Norm ();
        writer . SetLocAndOrientation (origin, normal, over);
        return true;
      }
    return false;
  }
};

static const char* DEFAULT_POOL = "leap";

void Setup ()
{ const char *leap_pool = getenv ("LEAP_POOL");
  Splash *splash = new Splash (NULL == leap_pool ?
                               DEFAULT_POOL : leap_pool);
  bool loaded = false;
  if (1 < args . Count ())
    { Protein p = LoadProtein (args . Nth (1));
      OB_LOG_WARNING ("Configuring leap from: %s", args . Nth (1) . utf8 ());
      if (! p . IsNull ())
        loaded = splash -> Configure (p);
    }
  if (! loaded)
    { Protein p = LoadProtein ("/etc/oblong/screen.protein");
      OB_LOG_WARNING ("/etc/oblong/screen.protein");
      if (! p . IsNull ())
        loaded = splash -> Configure (p);
    }
  if (! loaded)
    OB_LOG_WARNING ("Could not find or infer configuration information for "
                    "your LeapMotion");
}


splash.C

/* (c)  oblong industries */

// # splash
//
// Help your Leap Motion sensor dive in to pools

#include <Leap.h>
#define WORKER
#include <Greenhouse.h>
#include <list>
#include <boost/thread/mutex.hpp>

/// FrameWriter: A handy class to transform leap types into plasma types
class FrameWriter {
protected:
  Leap::Vector origin, normal, over;
  Leap::Matrix transform;
  Str provenance;

  /// Transform g-speak Vect into Leap Vector
  const Leap::Vector VV (Vect const& v) const
  { return Leap::Vector (v.x, v.y, v.z); }

  /// Takes a relative direction from Leap and returns it's absolute direction
  Leap::Vector const Direction (Leap::Vector const& v) const
  { return transform . transformDirection (v); }
  /// Takes a relative Location from Leap and returns it's absolute location
  Leap::Vector const Point (Leap::Vector const& v) const
  { return transform . transformPoint (v); }

  /// Transforms a Leap Vector in to a g-speak Vect and wraps it in a Slaw
  Slaw ToSlaw (Leap::Vector const &v) const
  { return Slaw (Vect (v.x, v.y, v.z)); }

  /// Transforms a list into a Slaw list
  template <typename Lst>
  Slaw ToSlaw (Lst const &ll) const {
    Slaw out = Slaw::List ();
    for (auto &x : ll)
      out = out . ListAppend (ToSlaw (x));
    return out;
  }

  /// Creates a slaw that fully describes a Leap Pointing gesture
  Slaw ToSlaw (Leap::Pointable const& p) const
  { Leap::Vector
      phys_origin = Point (p . tipPosition ()),
      phys_through = phys_origin + p . length () * Direction (p . direction ());
    return Slaw::Map ("id", p . id (),
                      "dir", ToSlaw (Direction (p . direction ())),
                      "hand", p . hand () . id (),
                      "isfngr", p . isFinger (),
                      "istool", p . isTool (),
                      "length", p . length ())
      . MapMerge (Slaw::Map ("t-pos", ToSlaw (phys_origin),
                             "t-vel", ToSlaw (p . tipVelocity ()),
                             "width", p . width (),
                             "orig", ToSlaw (phys_origin),
                             "thru", ToSlaw (phys_through)));
  }

  /// Creates a slaw that fully describes a Hand found by the Leap
  Slaw ToSlaw (Leap::Hand const& h) const
  { Leap::Vector
      phys_origin  = Point (h . palmPosition ()),
      phys_through = phys_origin + Direction (h . direction ());
    return Slaw::Map ("id", h . id (),
                      "dir", ToSlaw (Direction (h . direction ())),
                      "plmnrm", ToSlaw (Direction (h . palmNormal ())),
                      "plmpos", ToSlaw (phys_origin),
                      "plmvel", ToSlaw (h . palmVelocity ()),
                      "center", ToSlaw (Direction (h . sphereCenter ())),
                      "radius", h . sphereRadius (),
                      "orig", ToSlaw (phys_origin),
                      "thru", ToSlaw (phys_through));
  }

  /// Returns a Leap::Gesture::State as a slaw wrapped string
  Slaw ToSlaw (Leap::Gesture::State const& s) const
  { switch (s) {
    case Leap::Gesture::STATE_START: return Slaw ("start");
    case Leap::Gesture::STATE_UPDATE: return Slaw ("update");
    case Leap::Gesture::STATE_STOP: return Slaw ("stop");
    default: return Slaw ("invalid");
    }
  }


  /// Transforms a list into a Slaw list of just id's
  template <typename Lst>
  Slaw IDs (Lst const& lst) const
  { Slaw out = Slaw::List ();
    for (auto &x : lst)
      out = out . ListAppend (x . id ());
    return out;
  }

  /// Creates a slaw that fully describes a Leap Gesture
  Slaw ToSlaw (Leap::Gesture const& g) const
  { Slaw out = Slaw::Map ("id", g . id (),
                          "dur", g . duration (),
                          "dursec", g . durationSeconds (),
                          "hands", IDs (g . hands ()),
                          "pntrs", IDs (g . pointables ()),
                          "state", ToSlaw (g . state ()));
    switch (g . type ()) {
    case Leap::Gesture::TYPE_SWIPE:
      return ToSlaw ((Leap::SwipeGesture const&) g, out);
    case Leap::Gesture::TYPE_CIRCLE:
      return ToSlaw ((Leap::CircleGesture const&) g, out);
    case Leap::Gesture::TYPE_SCREEN_TAP:
      return ToSlaw ((Leap::ScreenTapGesture const&) g, out);
    case Leap::Gesture::TYPE_KEY_TAP:
      return ToSlaw ((Leap::KeyTapGesture const&) g, out);
    default:
      out = out . MapPut ("type", "invalid");
      return out;
    }
  }

  /// Creates a slaw that fully describes a Leap "circle" gesture
  Slaw ToSlaw (Leap::CircleGesture const& c, Slaw out) const
  { return out . MapMerge
      (Slaw::Map ("type", "circle",
                  "point", c . pointable () . id (),
                  "cent", ToSlaw (Point (c . center ())),
                  "norm", ToSlaw (Direction (c . normal ())),
                  "prog", c . progress (),
                  "radius", c . radius ()));
  }

  /// Creates a slaw that fully describes a Leap "swipe" gesture
  Slaw ToSlaw (Leap::SwipeGesture const& s, Slaw out) const
  { return out . MapMerge
      (Slaw::Map ("type", "swipe",
                  "dir", ToSlaw (Direction (s . direction ())),
                  "point", s . pointable () . id (),
                  "pos", ToSlaw (Point (s . position ())),
                  "speed", s . speed (),
                  "start", ToSlaw (Point (s . startPosition ()))));
  }

  /// Creates a slaw that fully describes a Leap "screen tap" gesture
  Slaw ToSlaw (Leap::ScreenTapGesture const& s, Slaw out) const
  { return out . MapMerge
      (Slaw::Map ("type", "s-tap",
                  "dir", ToSlaw (Direction (s . direction ())),
                  "point", s . pointable () . id (),
                  "pos", ToSlaw (Point (s . position ())),
                  "prog", s . progress ()));
  }

  /// Creates a slaw that fully describes a Leap "key tap" gesture
  Slaw ToSlaw (Leap::KeyTapGesture const& s, Slaw out) const
  { return out . MapMerge
      (Slaw::Map ("type", "k-tap",
                  "dir", ToSlaw (Direction (s . direction ())),
                  "point", s . pointable () . id (),
                  "pos", ToSlaw (Point (s . position ())),
                  "prog", s . progress ()));
  }

  /// returns the hostname of the machine running splash
  static const Str Hostname ()
  { const size_t MAX_HOSTNAME = 64;
    char hostname[MAX_HOSTNAME];
    if (-1 != gethostname (hostname, MAX_HOSTNAME))
      return Str (hostname);
    else
      return Str ();
  }

  /// creates a provenance that uniquely identifies the leap connected to the
  /// machine running splash using the hostname
  static const Str DefaultProvenance ()
  { return "leap-" + Hostname (); }

public:
  FrameWriter ()  :  transform (Leap::Matrix::identity ()),
                     provenance (DefaultProvenance ()) {}

  void SetProvenance (Str const &p) { provenance = p; }

  /// Sets up the location and orientation of the leap in order to properly
  /// transform relatively located leap events into absolutely located ones.
  void SetLocAndOrientation (Vect const& orig, Vect const& nrm, Vect const& ov)
  { origin = VV (orig);
    normal = VV(nrm);
    over = VV(ov);
    transform = Leap::Matrix (over, normal, over . cross (normal), origin);
  }

  /// Takes an event/frame generated from a leap and returns a Slaw that fully
  /// describes that event
  Slaw ToSlaw (Leap::Frame const& frame) const
  { return Slaw::Map ("ts", frame . timestamp (),
                      "id", frame . id (),
                      "hands", ToSlaw (frame . hands ()),
                      "pntrs", ToSlaw (frame . pointables ()),
                      "gests", ToSlaw (frame . gestures ()));
  }

  /// Returns a slaw the provides the leap's location, orientation and name
  Slaw DeviceDescription () const
  { return Slaw::Map ("orig", ToSlaw (origin),
                      "norm", ToSlaw (normal),
                      "over", ToSlaw (over),
                      "prov", provenance);
  }
};

/// Callback handler for Leap events
template <typename Callback>
class SplashListener  :  public Leap::Listener {
protected:
  Callback &callback;
public:
  SplashListener (Callback &c)
    : callback (c) {}
  virtual void onFrame (const Leap::Controller &c)
  { callback . DepositFrame (c . frame ()); }
};

// Surely there's a way to get this from Leap's API...?
static const Str LEAP_VERSION = "0.7.9";
static const float64 DEFAULT_Z_DISTANCE = 500.0;
static const float64 DEFAULT_Y_DISTANCE = 200.0;

/// The main class type for Splash. It sets up the leap controller, the
/// callback handler for the controller and handles each frame using
/// the FrameWriter class to transform a leap frame into a g-speak protein
/// and deposits the protein into the designated output pool.
class Splash  :  public Thing {
private:
  Str const pool;
  SplashListener <Splash> listener;
  Leap::Controller controller;
  boost::mutex mutex;
  FrameWriter writer;
  std::list<Protein> to_deposit;
  bool stop_depositing;
public:
  Splash (Str const& output_pool)
    :  Thing (),
       pool (output_pool),
       listener (*this),
       controller (listener),
       stop_depositing (false)
  { ParticipateInPool (pool);
    controller . enableGesture (Leap::Gesture::TYPE_CIRCLE);
    controller . enableGesture (Leap::Gesture::TYPE_SWIPE);
    controller . enableGesture (Leap::Gesture::TYPE_SCREEN_TAP);
    controller . enableGesture (Leap::Gesture::TYPE_KEY_TAP);
    controller . setPolicyFlags (Leap::Controller::POLICY_BACKGROUND_FRAMES);
  }

  virtual ~Splash ()
  { // Make sure we don't try to deposit any more proteins after
    // the application shuts down.
    { boost::mutex::scoped_lock lck (mutex);
      stop_depositing = true;
    }
    if (controller . isConnected ())
      controller . removeListener (listener);
  }

  /// The handler function for each Leap frame, it takes a leap frame,
  /// transforms it to a protein and deposits it to the leap pool
  void DepositFrame (Leap::Frame const& f)
  { Protein p (Slaw::List ("greenhouse", "leap", LEAP_VERSION),
               Slaw::Map ("leap", writer . DeviceDescription (),
                          "frame", writer . ToSlaw (f)));

    // The leap runs in its own thread.  Make sure that we don't try
    // to deposit two proteins at the same time.
    boost::mutex::scoped_lock lck (mutex);
    if (! stop_depositing)
      to_deposit . push_back (p);
  }

  virtual void Travail ()
  { std::list<Protein> copy;
    { boost::mutex::scoped_lock lck (mutex);
      std::swap (copy, to_deposit);
    }
    for (Protein const& p : copy)
      Deposit (p, pool);
  }

  /// Reads in a configuration protein to set up the spatial layout of
  /// the leap
  bool Configure (Protein const& p)
  { Slaw ing = p . Ingests ();
    Vect origin, normal, over;
    Str prov;
    Slaw leap = ing . MapFind ("leap");
    if (leap . MapFind ("provenance") . Into (prov))
      writer . SetProvenance (prov);

    if (leap . MapFind ("cent") . Into (origin) &&
        leap . MapFind ("norm") . Into (normal) &&
        leap . MapFind ("over") . Into (over))
      { writer . SetLocAndOrientation (origin, normal, over);
        return true;
      }

    // If there's no explicit Leap information, infer it from
    // the screen protein, using the ``main'' screen.  We'll
    // assume that the leap is about 50cm back and 20 cm down
    // from the center of the screen and is pointing up.
    Slaw screen = ing . MapFind ("screens") . MapFind ("main");
    if (screen . MapFind ("cent") . Into (origin) &&
        screen . MapFind ("over") . Into (over))
      { normal = Vect (0, 1, 0);
        origin -= DEFAULT_Y_DISTANCE * normal . Norm ()
          + DEFAULT_Z_DISTANCE * normal . Cross (over) . Norm ();
        writer . SetLocAndOrientation (origin, normal, over);
        return true;
      }
    return false;
  }
};

static const char* DEFAULT_POOL = "leap";

void Setup ()
{ const char *leap_pool = getenv ("LEAP_POOL");
  Splash *splash = new Splash (NULL == leap_pool ?
                               DEFAULT_POOL : leap_pool);
  bool loaded = false;
  if (0 < args . Count ())
    { Protein p = LoadProtein (args [0]);
      if (! p . IsNull ())
        loaded = splash -> Configure (p);
    }
  if (! loaded)
    { Protein p = LoadProtein ("/etc/oblong/screen.protein");
      if (! p . IsNull ())
        loaded = splash -> Configure (p);
    }
  if (! loaded)
    WARN ("Could not find or infer configuration information for your LeapMotion");
}
